_posts/2017-06-08-Rust-1.18.md

```diff
@@ -0,0 +1,200 @@
+size of six bytes. But looking at the source, you'd expect it to have four. The
+extra two bytes come from padding; given that we have a `u16` here, it should be
+aligned to two bytes. But in this case, it's at offset one. To move it to offset
+two, another byte of padding is placed after the first `u8`. To give the whole struct
+a proper alignment, another byte is added after the second `u8` as well, giving us
+`1 + 1 (padding) + 2 + 1 + 1 (padding) = 6 bytes`.
+
+But what if our struct looked like this?
+
+```rust
+struct Optimal(u8, u8, u16);
+```
+
+This struct is properly aligned; the `u16` lies on a two byte boundary, and so does
+the entire struct. No padding is needed. This gives us `1 + 1 + 2 = 4 bytes`.
+
+When designing Rust, stuff like this is why we left the details here undefined.
+It allows for exactly this kind of thing: the compiler can optimize `Suboptimal`
+into `Optimal` automatically. And if you check the sizes of `Suboptimal` and
+`Optimal` on Rust 1.18.0, you'll see that they both have a size of four bytes.
+
+We've been planning this for a while; previous versions of Rust included this
+optimization on the nightly channel, but some people wrote unsafe code that
+assumed the exact details of the representation. We rolled it back while we fixed
+all instances of this that we know about, but if you find some code breaks due to
+this, you should fix it! In the meantime, there's also a flag you can use to control
+this. Imagine we have `Suboptimal` in a file named `foo.rs`, along with a `main`
+function that prints out its size:
+
+```rust
+> rustc foo.rs
+> ./foo
+4
+> rustc foo.rs -Z fuel=foo=0
+optimization-fuel-exhausted: Reorder fields of "Suboptimal"
+> ./foo
+6
+```
+
+This flag is based on an idea called "[Optimization
+fuel](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/dfopt-popl10.pdf)":
+
+> It works by giving the optimizer a finite supply of optimization fuel. Each
+> time a rewrite function proposes to replace a node, one unit of fuel is
+> consumed. When the optimizer runs out of fuel, further rewrites are
+> suppressed.
+
+Currently, this is the only optimization that uses the "fuel" concept. By
+setting the fuel to zero, `rustc` will not perform the optimization. This can
+keep your code compiling if you run into a problem, but the longer-term fix is
+to use a `#[repr]` attribute to guarantee a particular layout if you rely on it.
+
+We've been planning on moving `rustdoc` to use a CommonMark compliant markdown
+parser for a long time now. However, just switching over can introduce
+regressions where the CommonMark spec differs from our existing parser, Hoedown.
+As part of the transition plan, [a new flag has been added to
+`rustdoc`](https://github.com/rust-lang/rust/pull/40338), `--enable-commonmark`.
+This will use the new parser instead of the old one. Please give it a try! There's
+no scenario we know of where tweaking your markdown gets identical results on both
```

Comment:
. . . . is this sentence supposed to say "There's no scenario we know of where tweaking your markdown *doesn't* get identical results on both parsers...?" If not, I'm confused about why we're stating this in this way? Does "tweaking your markdown" mean I change the markdown text I've written, or does it mean "tweaking your markdown parser"?